### PR TITLE
Filmic HDR tone mapping

### PR DESCRIPTION
**Pull request by RealBadAngel**

### CODE REVIEW

**Comment by kaadmy** [comment]
:thumbsup:

**Comment by RealBadAngel** [comment]
screenshot_20160125_182704
screenshot_20160125_182728

**Comment by RealBadAngel** [comment]
screenshot_20160125_183100
screenshot_20160125_183129

**Comment by RealBadAngel** [comment]
screenshot_20160125_183545
screenshot_20160125_183606

**Comment by Fixer-007** [comment]
:+1:

**Comment by paramat** [comment]
No objections from me, as this seems to not affect the game when shaders are disabled.

**Comment by RealBadAngel** [comment]
screenshot_20160126_031120
screenshot_20160126_031156

**Comment by RealBadAngel** [comment]
screenshot_20160126_033903
screenshot_20160126_033923

**Comment by paramat** [comment]
^
After
Before
.. then ..
Before
After
?

**Comment by RealBadAngel** [comment]
@maybe ive mixed the order, but definitely you can say on your own which one is after in any case.
More juicy, more saturated ones. Please do note the last two pairs are made with tweaked filter (before i was messing with it but decided to go back to default filter settings)

**Comment by C1ffisme** [comment]
Nice!

:+1:

**Comment by paramat** [comment]
The colours seem over-saturated now, the HDR effect seems reversed, darkplaces are darker and bright things brighter, i thought HDR did the opposite. However i do need to actually test this in-game to get a better impression.

**Comment by RealBadAngel** [comment]
@paramat 
"The method of rendering an HDR image to a standard monitor or printing device is called tone mapping. This method reduces the overall contrast of an HDR image to facilitate display on devices or printouts with lower dynamic range, and can be applied to produce images with preserved or exaggerated local contrast for artistic effect."

File: client/shaders/nodes_shader/opengl_fragment.glsl [hunk 1]
```diff
@@ -20,6 +20,32 @@ bool normalTexturePresent = false;
...
+vec3 uncharted2Tonemap(vec3 x)
+{
+	return ((x * (A * x + C * B) + D * E) / (x * (A * x + B) +D * F)) - E / F;
+}
+
+vec4 applyToneMapping(vec4 color)
+{
+	color = vec4(pow(color.rgb, vec3(2.2)), color.a);
+	const float gamma = 1.6;
+    float exposureBias = 5.5;
```
Comment (nerzhul) [inline client/shaders/nodes_shader/opengl_fragment.glsl]: please use consistent indentation

**Comment by RealBadAngel** [comment]
@nerzhul fixed

**Comment by ObaniGemini** [comment]
:+1: :+1:

**Comment by sfan5** [comment]
:+1:

**Comment by ObaniGemini** [comment]
Merge ?

File: client/shaders/nodes_shader/opengl_fragment.glsl [hunk 2]
```diff
@@ -160,22 +186,27 @@ void main(void)
 	color = base.rgb;
 #endif
 
+	vec4 col = vec4(color.rgb, 1.0);
```
Comment (nerzhul) [inline client/shaders/nodes_shader/opengl_fragment.glsl]: can you merge this calcul with the next line to limit calculs and permit GPU to optimize this calcul ?

File: client/shaders/water_surface_shader/opengl_fragment.glsl [hunk 3]
```diff
@@ -116,22 +142,27 @@ vec4 base = texture2D(baseTexture, uv).rgba;
 	color = base.rgb;
 #endif
 
+	vec4 col = vec4(color.rgb, 1.0);
```
Comment (nerzhul) [inline client/shaders/water_surface_shader/opengl_fragment.glsl]: same here

**Comment by nerzhul** [comment]
please add the previous optimization. After this optimization :+1: for merge

**Comment by paramat** [comment]
No objections as i understand this is using industry default settings not RBA's settings.

**Comment by paramat** [comment]
What method and defaults are used? Reinhard, Hable or other?

**Comment by RealBadAngel** [comment]
@paramat its Hable's UC2 filmic

**Comment by Fixer-007** [comment]
Crash to desktop if minetest.conf does not have enable_tonemapping setting.
2016-02-08 15:35:57: ERROR[Main]: Some exception: "Setting [enable_tonemapping] not found."

**Comment by kilbith** [comment]
Tested. I don't see that as an improvement, honestly.

The only thing it's doing significantly is to make the game brighter and/or over-saturated.

**Comment by RealBadAngel** [comment]
@kilbith color perception is different for everybody, what you said "the only thing it's doing..." is just the thing its supposed to be doing

**Comment by RealBadAngel** [comment]
@Fixer-007 fixed

**Comment by paramat** [comment]
Added line comments.

**Comment by paramat** [comment]
I don't fully understand the code but i generally approve, with some research i find Hable's UC2 one of the better-looking methods.

File: builtin/settingtypes.txt [hunk 4]
```diff
@@ -323,6 +323,11 @@ fsaa (FSAA) enum 0 0,1,2,4,8,16
 #    Thy only work with the OpenGL video backend.
 enable_shaders (Shaders) bool true
 
+[****Tone Mapping]
+
+#    Enables filmic tone mapping
```
Comment (kwolekr) [inline builtin/settingtypes.txt]: Could you add a note somewhere that this effect requires shaders to be enabled?

File: client/shaders/nodes_shader/opengl_fragment.glsl [hunk 5]
```diff
@@ -20,6 +20,32 @@ bool normalTexturePresent = false;
...
+const float B = 0.30;
+const float C = 0.10;
+const float D = 0.20;
+const float E = 0.01;
+const float F = 0.30;
+const float W = 11.2;
+
+vec3 uncharted2Tonemap(vec3 x)
+{
+	return ((x * (A * x + C * B) + D * E) / (x * (A * x + B) + D * F)) - E / F;
```
Comment (kwolekr) [inline client/shaders/nodes_shader/opengl_fragment.glsl]: ..I don't suppose there's any way for you to describe the black magic going on here, is there?

File: src/shader.cpp [hunk 6]
```diff
@@ -764,6 +764,9 @@ ShaderInfo generate_shader(std::string name, u8 material_type, u8 drawtype,
 	else
 		shaders_header += "0\n";
 
+	if (g_settings->getBool("tone_mapping"))
+		shaders_header += "#define ENABLE_TONE_MAPPING\n";
```
Comment (kwolekr) [inline src/shader.cpp]: Doesn't this have to be #define ENABLE_TONE_MAPPING 1 for ATI shaders?

File: client/shaders/water_surface_shader/opengl_fragment.glsl [hunk 7]
```diff
@@ -22,6 +22,32 @@ bool texSeamless = false;
...
+vec3 uncharted2Tonemap(vec3 x)
+{
+	return ((x * (A * x + C * B) + D * E) / (x * (A * x + B) + D * F)) - E / F;
+}
+
+vec4 applyToneMapping(vec4 color)
+{
+	color = vec4(pow(color.rgb, vec3(2.2)), color.a);
+	const float gamma = 1.6;
+	float exposureBias = 5.5;
```
Comment (kwolekr) [inline client/shaders/water_surface_shader/opengl_fragment.glsl]: shouldn't exposureBias be a const float as well?

File: client/shaders/water_surface_shader/opengl_fragment.glsl [hunk 8]
```diff
@@ -22,6 +22,32 @@ bool texSeamless = false;
...
+	color.rgb = uncharted2Tonemap(exposureBias * color.rgb);
+	vec3 whiteScale = 1.0 / uncharted2Tonemap(vec3(W, W, W));
```
Comment (kwolekr) [inline client/shaders/water_surface_shader/opengl_fragment.glsl]: are you able to cache the result of this line somehow?

File: src/shader.cpp (hunk 6 above)
Comment (RealBadAngel) [inline src/shader.cpp]: No it doesn't, its perfectly fine

Comment (nerzhul) [inline src/shader.cpp]: @kwolekr i will test it this evening and see if this works, i have an AMD GPU

**Comment by kwolekr** [comment]
https://github.com/minetest/minetest/commit/eb3840a3f88dd9902ff1c5ea50d90a7f33e08d82

**Pull request closed without merging**